### ISSUE TITLE
Kubernetes deployment not working

### ISSUE THREAD

**Issue by Smidra** [DescartesResearch/TeaStore#140]
Hi, I am planning to use TeaStore for my Bachelor Thesis but it looks like the kubernetes .yaml config is outdated.

When I try to 
`kubectl create -f https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml`

My kubernetes cluster (using kind + latest k8s) fails with:

```
service/teastore-db created
service/teastore-registry created
service/teastore-persistence created
service/teastore-auth created
service/teastore-image created
service/teastore-recommender created
service/teastore-webui created
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
unable to recognize "https://raw.githubusercontent.com/DescartesResearch/TeaStore/master/examples/kubernetes/teastore-clusterip.yaml": no matches for kind "Deployment" in version "extensions/v1beta1"
```

I googled the bug and found a solution here:
https://stackoverflow.com/questions/58481850/no-matches-for-kind-deployment-in-version-extensions-v1beta1

So i downloaded the .yaml and replaced "extensions/v1beta1" with "apps/v1"

I tried it again with:

`kubectl create -f teastore-new.yaml`

But it fails with "missing selector"

```
error: error validating "teastore-new.yaml": error validating data: ValidationError(Deployment.spec): missing required field "selector" in io.k8s.api.apps.v1.DeploymentSpec; if you choose to ignore these errors, turn validation off with --validate=false
```

If we turn validation off with  "--validate=false" we get an error 

```
service/teastore-db created
service/teastore-registry created
service/teastore-persistence created
service/teastore-auth created
service/teastore-image created
service/teastore-recommender created
service/teastore-webui created
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-db" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-db"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-registry" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-registry"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-persistence" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-persistence"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-auth" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-auth"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-image" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-image"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-recommender" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-recommender"}: `selector` does not match template `labels`]
Error from server (Invalid): error when creating "teastore-new.yaml": Deployment.apps "teastore-webui" is invalid: [spec.selector: Required value, spec.template.metadata.labels: Invalid value: map[string]string{"app":"teastore", "run":"teastore-webui"}: `selector` does not match template `labels`]
```

It looks like the yaml files are outdated are in need of and update (they need a selector), please could you look into it?
Thanks a lot, 
Smidra

**Comment by Smidra** [comment]
So, I've stiched together a new, wierd **teastore-clusterip.yaml**  with selectors. It "works" but I am a total newbie in k8s so I would really appriciate If someone took a look at it. :D 

I think i messed something up, because when I run it the Teastore app does not recognise how many replicas of each MS are running. (# in status site)

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-db
spec:
  selector:
    matchLabels:
      app: teastore-db
  template:
    metadata:
      labels:
        app: teastore-db
        run: teastore-db
    spec:
      containers:
      - name: teastore-db
        image: descartesresearch/teastore-db
        ports:
        - containerPort: 3306
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-db
  labels:
    app: teastore-db
    run: teastore-db
spec:
  selector:
    app: teastore-db
  ports:
  - port: 3306
    protocol: TCP
  selector:
    run: teastore-db
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-registry
spec:
  selector:
    matchLabels:
      app: teastore-registry
  template:
    metadata:
      labels:
        app: teastore-registry
        run: teastore-registry
    spec:
      containers:
      - name: teastore-registry
        image: descartesresearch/teastore-registry
        ports:
        - containerPort: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-registry
  labels:
    app: teastore-registry
    run: teastore-registry
spec:
  selector:
    app: teastore-registry
  ports:
  - port: 8080
    protocol: TCP
  selector:
    run: teastore-registry
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-persistence
spec:
  selector:
    matchLabels:
      app: teastore-persistence
  template:
    metadata:
      labels:
        app: teastore-persistence
        run: teastore-persistence
    spec:
      containers:
      - name: teastore-persistence
        image: descartesresearch/teastore-persistence
        ports:
        - containerPort: 8080
        env:
        - name: HOST_NAME
          value: "teastore-persistence"
        - name: REGISTRY_HOST
          value: "teastore-registry"
        - name: DB_HOST
          value: "teastore-db"
        - name: DB_PORT
          value: "3306"
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-persistence
  labels:
    app: teastore-persistence
    run: teastore-persistence
spec:
  selector:
    app: teastore-persistence
  ports:
  - port: 8080
    protocol: TCP
  selector:
    run: teastore-persistence
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-auth
spec:
  selector:
    matchLabels:
      app: teastore-auth
  template:
    metadata:
      labels:
        app: teastore-auth
        run: teastore-auth
    spec:
      containers:
      - name: teastore-auth
        image: descartesresearch/teastore-auth
        ports:
        - containerPort: 8080
        env:
        - name: HOST_NAME
          value: "teastore-auth"
        - name: REGISTRY_HOST
          value: "teastore-registry"
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-auth
  labels:
    app: teastore-auth
    run: teastore-auth
spec:
  selector:
    app: teastore-auth
  ports:
  - port: 8080
    protocol: TCP
  selector:
    run: teastore-auth
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-image
spec:
  selector:
    matchLabels:
      app: teastore-image
  template:
    metadata:
      labels:
        app: teastore-image
        run: teastore-image
    spec:
      containers:
      - name: teastore-image
        image: descartesresearch/teastore-image
        ports:
        - containerPort: 8080
        env:
        - name: HOST_NAME
          value: "teastore-image"
        - name: REGISTRY_HOST
          value: "teastore-registry"
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-image
  labels:
    app: teastore-image
    run: teastore-image
spec:
  selector:
    app: teastore-image
  ports:
  - port: 8080
    protocol: TCP
  selector:
    run: teastore-image
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-recommender
spec:
  selector:
    matchLabels:
      app: teastore-recommender
  template:
    metadata:
      labels:
        app: teastore-recommender
        run: teastore-recommender
    spec:
      containers:
      - name: teastore-recommender
        image: descartesresearch/teastore-recommender
        ports:
        - containerPort: 8080
        env:
        - name: HOST_NAME
          value: "teastore-recommender"
        - name: REGISTRY_HOST
          value: "teastore-registry"
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-recommender
  labels:
    app: teastore-recommender
    run: teastore-recommender
spec:
  selector:
    app: teastore-recommender
  ports:
  - port: 8080
    protocol: TCP
  selector:
    run: teastore-recommender
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: teastore-webui
spec:
  selector:
    matchLabels:
      app: teastore-webui
  template:
    metadata:
      labels:
        app: teastore-webui
        run: teastore-webui
    spec:
      containers:
      - name: teastore-webui
        image: descartesresearch/teastore-webui
        ports:
        - containerPort: 8080
        env:
        - name: HOST_NAME
          value: "teastore-webui"
        - name: REGISTRY_HOST
          value: "teastore-registry"
---
apiVersion: v1
kind: Service
metadata:
  name: teastore-webui
  labels:
    app: teastore-webui
    run: teastore-webui
spec:
  selector:
    app: teastore-webui
  type: NodePort
  ports:
  - port: 8080
    nodePort: 30080
    protocol: TCP
  selector:
    run: teastore-webui
```

**Comment by SimonEismann** [comment]
I just tested this with Kubernetes version 1.15.9 (the newest version available on GKE right now, aws still uses 1.14.X) and was not able to reproduce this issue. I'll try to reproduce it with Kubernetes version 1.17 in the next couple of days.

Until then, could you try to deploy a slightly older version of Kubernetes (kind seems to support many versions) and see if the deployment works then?

**Comment by SimonEismann** [comment]
Based on your initial work, in pull requests #148/#149 we added new kubernetes deployment descriptors for kubernetes version 1.16 onward labeled as _1.16 in the examples folder. Let me know if these don't fix the issue.
